services/code_hosting/git_hub_service.rb

```diff
@@ -1,118 +1,352 @@
+            if key.nil?
+              fastfile_config[:no_platform] = fastfile.tree[key]
+            else
+              fastfile_config[key.to_sym] = fastfile.tree[key]
+            end
+          end
+          fastfile_json_path = File.join(path, "fastfile.json")
+          FileUtils.touch(fastfile_json_path) unless File.exist?(fastfile_json_path) && !File.zero?(fastfile_json_path)
+          File.write(fastfile_json_path, JSON.pretty_generate(fastfile_config))
+          self.cache[repo + branch] = fastfile_config
+          return fastfile_config
+        rescue ArgumentError
+          self.setup_auth(
+            repo_url: repo_url,
+            provider_credential: provider_credential,
+            path: path
+          )
+          self.clone(
+            repo_url: repo_url,
+            branch: branch,
+            provider_credential: provider_credential
+          )
+          self.peek_fastfile_configuration(
+            repo_url: repo_url,
+            branch: branch,
+            provider_credential: provider_credential
+          )
+          self.unset_auth
+        end
+      end
+
+      # Class method that shallow clones the repo on the target branch.
+      # @param [String] repo_url
+      # @param [String] branch
+      # @param [GithubProviderCredential] provider_credential
+      # @return [Git::Base]
+      def clone(repo_url: nil, branch: "master", provider_credential: nil, path: self.temp_path)
+        repo = self.repo_from_url(repo_url)
+        path = File.join(path, repo, branch)
+        FileUtils.rm_rf(path) if File.directory?(path)
+        FileUtils.mkdir_p(path)
+        self.setup_auth(repo_url: repo_url, provider_credential: provider_credential, path: path)
+        Git.clone(url_from_repo(repo_url), repo.split("/").last,
+                  path: path,
+                  recursive: true,
+                  depth: 1)
+        git = Git.open(File.join(path, repo.split("/").last))
+        git.branch(branch).checkout
+        return git
+      end
 
-    GitHubService.status_context_prefix = "fastlane.ci: "
+      # Class method that finds the directory for the first Fastfile found given a root_path
+      # @param [String] root_path
+      # @return [String] the path of the Fastfile
+      def fastfile_path(root_path: nil)
+        fastfiles = Dir[File.join(root_path, "fastlane/Fastfile")]
+        fastfiles = Dir[File.join(root_path, "**/fastlane/Fastfile")] if fastfiles.count == 0
+        fastfile_path = fastfiles&.first
+        return fastfile_path
+      end
+
+      # Class method that setups the authentication needed for making git operations.
+      # @param [String] repo_url
+      # @param [GithubProviderCredential] provider_credential
+      # @param [String] path
+      def setup_auth(repo_url: nil, provider_credential: nil, path: nil)
+        repo = repo_from_url(repo_url)
+        self.temporary_storage_path = File.join(self.temporary_git_storage, "git-auth-#{SecureRandom.uuid}")
```

Comment:
BTW, I've seen this https://github.com/fastlane/ci/pull/290 @taquitos, should I implement this as you've added in that PR? What that tries to achieve? to remove the duplication of the temporary_auths being created for each clone?

Reply:
Yes, you should definitely take that change I made. We don't need tons of unique temporary auths being created. We just need 1 temp per checkout, so that's what my update did 😄

Reply:
Gonna steal that for the next commit 😄